File: java/L500/LeetCode0503.java
package L500;

import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;

public class LeetCode0503 {
    public int[] nextGreaterElements(int[] nums) {
        final int N = nums.length;
        int[] res = new int[N];
        Arrays.fill(res, -1);
        Deque<Integer> stack = new LinkedList<>();
        for (int i = 0;i < 2 * N - 1;++i) {
            while (!stack.isEmpty() && nums[i % N] > nums[stack.getFirst() % N]) {
                int index = stack.removeFirst();
                res[index % N] = nums[i % N];
            }
            stack.addFirst(i);
        }
        return res;
    }
}

File: cpp/L500/LeetCode0503.cc
#include <stack>
#include <vector>
using std::vector;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        std::stack<int> st;
        int N = nums.size();
        vector<int> ans(N, -1);
        for (int i{0};i < 2 * N;++i) {
            int x{nums[i % N]};
            while (!st.empty() && nums[st.top()] < x) {
                int id = st.top();
                st.pop();
                ans[id] = x;
            }
            st.push(i % N);
        }
        return ans;
    }
};